python.md

```diff
@@ -957,6 +958,28 @@ from typing import Optional, Tuple, Dict
 from collections import defaultdict
 ```
 
+### Avoid wildcard imports
+
+It's tempting to use wildcard imports (ie `from somewhere import *`) in
+`__init__.py` package modules, where each imported module declares an `__all__`
+module variable. However, this is less clear for maintainers and
+harder for static analysis tools (like mypy) to work out what is going on. It's
+better to explicitly import the required objects.
+
+So instead of:
+```py
+# hallandoates/__init__.py
+from ._problems import *  # noqa
+from ._features import *  # noqa
+```
+prefer:
+```py
+# hallandoates/__init__.py
+from ._problems import ICantGoForThat, NoCanDo
+from ._features import man_eater, rich_girl, shes_gone
```

Comment:
This would require a significant amount of work to bring the codebase in line with this convention - although I like the idea of having perfectly typed code, is it worth doing?

Reply:
I know we're currently seeing a problem related to wildcards and type checking. Maybe the behaviour we're currently seeing is a mypy bug. Would you still propose/support this convention if mypy was working?

Do we have any references for what mypy behaviour with wildcard imports and __all__ should be? I can see [PRs](https://github.com/PyCQA/pylint/pull/1321) [in mypy](https://github.com/python/mypy/pull/1640) indicating the they intend to support them.

Reply:
> I know we're currently seeing a problem related to wildcards and type checking. Maybe the behaviour we're currently seeing is a mypy bug. Would you still propose/support this convention if mypy was working?

I would, but I appreciate the improvement in maintainability is small. Hence, I would adopt this going forward and only drip-feed clean-up.